### PR TITLE
Change `AllSatisfy` to succeed on empty collections

### LINKED ISSUE #2169: `AllSatisfy()` fails on empty collections
### Discussed in https://github.com/fluentassertions/fluentassertions/discussions/2143

<div type='discussions-op-text'>

<sup>Originally posted by **ArneMancofi** March  8, 2023</sup>
### Description

`new object[0].Should().AllSatisfy(o => { });` fails with 
> Expected collection to contain only items satisfying the inspector, but collection is empty.

This does not implement the documented effect of `AllSatisfy`:
> Asserts that a collection contains only items which meet the criteria provided by the inspector

In particular, any empty collection should always pass this test.

The bug is probably due to two code lines in 

https://github.com/fluentassertions/fluentassertions/blob/c3b9a7072b790dea7ce16b2ddbef933f31f6aa1f/Src/FluentAssertions/Collections/GenericCollectionAssertions.cs#L2929-L2930

### Reproduction Steps

```csharp
// Arrange
var emptyArray = new object[0];
// Act
// Assert
emptyArray.Should().AllSatisfy(o => { });
```

### Expected behavior

The test should pass without throwing an exception.

### Actual behavior

The test fails with exception
> Expected collection to contain only items satisfying the inspector, but collection is empty.

### Regression?

This has probably always been this way.

### Known Workarounds

Code like this would be a workaround:
```csharp
if (collection.Any()) collection.Should().AllSatisfy(...);
```

However, reading the documentation of `.AllSatisfy()`, it should not be necessary.

### Configuration

FluentAssertions 6.9.0
.NET 6.0

### Other information

The [WikiPedia article on universal quantification](https://en.wikipedia.org/wiki/Universal_quantification) states that 
> By convention, the formula ∀x∈∅ P(x) is always true, regardless of the formula P(x)

so it is the documentated behaviour of `.AllSatisfy()` that is conventionally correct, and the implementation that is not.</div>

### PR DESCRIPTION
**Pull request by jnyrup**
This fixes #2169

## IMPORTANT 

* [ ] If the PR touches the public API, the changes have been approved in a separate issue with the "api-approved" label.
* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/develop/Tests/FluentAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [x] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/develop/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://fluentassertions.com/releases).
* [ ] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/develop/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/develop/AcceptApiChanges.sh).
* [ ] If the PR affects [the documentation](../tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).
    * [ ] Please also run `./build.sh --target spellcheck` or `.\build.ps1 --target spellcheck` before pushing and check the good outcome

### CODE REVIEW

File: docs/_pages/releases.md [hunk 1]
```diff
@@ -35,6 +35,7 @@ sidebar:
     * `BeLessOrEqualTo`: Use `BeLessThanOrEqualTo`
 * Removed the `DefaultValueFormatter.SpacesPerIndentionLevel` property which was added during the development of v6, but wasn't removed before the release of v6 - [#2281](https://github.com/fluentassertions/fluentassertions/pull/2281)
 * Dropped direct support for .NET Core 2.x and .NET Core 3.x - [#2302](https://github.com/fluentassertions/fluentassertions/pull/2302)
+* `AllSatisfy` now suceeds when asserting that an empty collection satifies some predicates - [#2321](https://github.com/fluentassertions/fluentassertions/pull/2321)
```
Comment (IT-VBFK) [inline docs/_pages/releases.md:38]:
```suggestion
* `AllSatisfy` now suceeds when asserting that an empty collection satisfies some predicates - [#2321](https://github.com/fluentassertions/fluentassertions/pull/2321)
```

**Review by dennisdoomen — APPROVED** [review]
What about the other methods asserting a collection?

**Comment by jnyrup** [comment]
> What about the other methods asserting a collection?

If I understand your question, it's about the other methods on `GenericCollectionAssertions` should accept empty subject collections?

The only other I could find where we could also change this is `OnlyContain`, which was also requested to be changed in #1492.

**Comment by IT-VBFK** [comment]
> The only other I could find where we could also change this is OnlyContain, which was also requested to be changed in https://github.com/fluentassertions/fluentassertions/issues/1492.

huh... in terms of consistency.. is this a desired change?

**Comment by jnyrup** [comment]
> > The only other I could find where we could also change this is `OnlyContain`, which was also requested to be changed in #1492.
> 
> huh... in terms of consistency.. is this a desired change?

Inconsistency between `AllSatisfy` and `OnlyContain`?

**Comment by IT-VBFK** [comment]
Haha 😂
No.. other way round: Is it desired to be changed?

**Comment by jnyrup** [comment]
To be sure I understand what you're asking.
Are you questioning whether _this_ PR should have been merged at all?

**Comment by IT-VBFK** [comment]
Ahm.. no 

Just asking if I can open a PR that aligns `OnlyContain` with `AllSatisfy`.

**Pull request merged**